src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,94 +10,202 @@ internal struct FileStatus
+        private Interop.Sys.FileStatus _fileCache;
+
+        // -1: if the file cache isn't initialized - Refresh should always change this value
+        //  0: if the file cache was initialized with no errors
+        // Positive number: the error code returned by the lstat call
+        private int _initializedFileCache;
+
         // The last cached stat information about the file
-        private Interop.Sys.FileStatus _fileStatus;
+        // Refresh only collects this if lstat determines the path is a symbolic link
+        private Interop.Sys.FileStatus _symlinkCache;
 
-        // -1 if _fileStatus isn't initialized, 0 if _fileStatus was initialized with no
-        // errors, or the errno error code.
-        private int _fileStatusInitialized;
+        // -1: if the symlink cache isn't initialized - Refresh only changes this value if lstat determines the path is a symbolic link
+        // 0: if the symlink cache was initialized with no errors
+        // Positive number: the error code returned by the stat call
+        private int _initializedSymlinkCache;
 
         // We track intent of creation to know whether or not we want to (1) create a
         // DirectoryInfo around this status struct or (2) actually are part of a DirectoryInfo.
-        internal bool InitiallyDirectory { get; private set; }
+        // Set to true during initialization when the DirectoryEntry's INodeType describes a directory
+        internal bool InitiallyDirectory { get; set; }
 
         // Is a directory as of the last refresh
-        internal bool _isDirectory;
+        // Its value can come from either the main path or the symbolic link path
+        private bool _isDirectory;
 
         // Exists as of the last refresh
         private bool _exists;
 
-        internal static void Initialize(
-            ref FileStatus status,
-            bool isDirectory)
+        private bool IsFileCacheInitialized    => _initializedFileCache == 0;
+        private bool IsSymlinkCacheInitialized => _initializedSymlinkCache == 0;
+
+        private bool AreCachesInitialized =>
+            // The file cache should always be successfully refreshed
+            _initializedFileCache == 0 &&
+            // The symbolic link cache only gets refreshed when the path is detected
+            // to be a symlink, so -1 is also acceptable
+            _initializedSymlinkCache <= 0;
+
+        // Check if the main path (without following symlinks) has the hidden attribute set
+        // Only use this if Refresh has been successfully called at least once
+        internal bool HasHiddenFlag
         {
-            status.InitiallyDirectory = isDirectory;
-            status._fileStatusInitialized = -1;
+            get
+            {
+                Debug.Assert(IsFileCacheInitialized);
+                return IsFileCacheInitialized &&
+                       (_fileCache.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN;
```

Comment:
If we have the assert why do need to check IsFileCacheInitialized in the return?

Reply:
I removed the assert and added a comment why we're not checking.